lib/vector-config/src/named.rs

```diff
@@ -0,0 +1,12 @@
+/// A component with a well-known name.
+///
+/// Users can derive this trait automatically by using the
+/// [`component_name`][vector_config::component_name] macro on their structs/enums.
+pub trait NamedComponent {
+    const NAME: &'static str;
+
+    /// Gets the name of the component.
+    fn get_component_name(&self) -> &'static str {
+        Self::NAME
+    }
```

Comment:
Why is there an associated constant _and_ an instance method on the trait, which returns the associated constant, you ask?

Using this in the future to power how `serde` names the enum variants for the big component enums will require being able to feed an expression that can be used as a string reference. Doing that with the associated constant itself -- `<StdinConfig as NamedComponent>::NAME` -- is the cleanest/most efficient way to do that.

However, we've also used `NamedComponent` to replace `SourceConfig::source_type`, which means we need it to be a base trait of `SourceConfig` so we can ensure it's always present. The problem arises, however: what should the value of `NAME` be when we implement `SourceConfig` for `Sources`?

Using `<Sources as NamedComponent>::NAME` wouldn't make any sense, because the component name is only relevant when you have a `Sources` value. This is where the instance method comes in, as it lets us forward to the underlying component's implementation of `NamedComponent` when we have a value that is `Sources`.

Additionally, this is the same reason why we can't do an instance-less version of the method (i.e. `fn get_component_name() -> &'string str { Self::NAME }`) because the result makes no sense for `Sources`, only the implementations on the component configuration types it wraps around.

In the future, if we can add support to `serde` to support it, we'll generate code that looks like `#[serde(rename_with = "<StdinConfig as NamedComponent>::NAME")]` for types that we know actually have a meaningful value for `NAME`, Everywhere else that was previously using `SourceConfig::source_type` just uses the instance method instead.